_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,133 @@
+	function f() public {
+		x = 1; // This write used to be removed by the yul optimizer.
+		g();
+		x = 2;
+	}
+	function g() internal {
+		// The important aspect of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which it terminates the transaction, since the state change above will be missing.
+		if (msg.data.length > 4) return;
+		assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via-IR with enabled optimizer will result ``f()`` without additional calldata to incorrectly return without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic store removal step, e.g. the following is almost
+identical, but unaffected, since ``g()`` will be inlined (in the snippet above the ``return;`` prevents easy inlining):
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1;
+		g();
+		x = 2;
+	}
+	function g() internal {
+		if (msg.data.length <= 4)
+			assembly { return(0,0) }
+	}
+}
+```
+
+
+## Technical Details
+
+On the Yul level the issue manifests if the following sequence occurs:
+
+- A storage write occurs (e.g. ``sstore(0, 1)``).
+- A user-defined Yul function is called and this function both has a control flow path terminates the transaction and a control flow path out of the function.
+- Another storage write occurs that makes the first storage write above redundant (e.g. ``sstore(0, 2)``).
+
+In this situation, the optimizer will treat the first storage write as globally redundant, despite the fact that the called Yul function may successfully terminate the transaction, in which case the first storage write is not redundant at all.
+
+The following is a full example of a vulnerable Yul snippet:
+
+```
+{
+  function f() {
+    if gt(calldatasize(), 4) { leave }
+    return(0,0)
+  }
+  sstore(0, 1) // This sstore will incorrectly be removed.
+  f()
+  sstore(0, 2)
+}
+```
+
+At the call of ``f()``, the first ``sstore`` should already be marked as ``used`` by the optimizer, since the control flow path of ``f()`` reaching the ``return(0,0)`` relies on it.
+However, the optimizer failed to mark it as such, resulting in the second ``sstore`` marking it as ``unused`` and staging it for removal.
+
+The issue only occurs in the presence of a call to a conditionally successfully terminating Yul function that survives the optimizer.
+The following similar looking snippet is not affected, since the optimizer will inline the call to ``f()`` before performing
+the buggy optimization step for the removal of storage writes.
+
+```
+{
+  function f() {
+    if iszero(lt(calldatasize(), 4)) { return(0,0) }
+  }
+  sstore(0, 1)
+  f()
+  sstore(0, 2)
+}
+```
+
+In this case, ``f()`` will be inlined, resulting in
+
+```
+{
+  sstore(0, 1)
+  if calldataload(4) { return(0,0) }
+  sstore(0, 2)
+}
+```
+
+In which both ``sstore``s are correctly retained.
```

Comment:
The description of the step would also make it clear why it has to be in a function to get triggered. I.e. that we're marking the functions as reverting/terminating/continuing.